qos/evm/state.go

```diff
@@ -0,0 +1,85 @@
+package evm
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/pokt-network/poktroll/pkg/polylog"
+
+	"github.com/buildwithgrove/path/protocol"
+)
+
+// ServiceState keeps the expected current state of the EVM blockchain based on the endpoints' responses to
+// different requests.
+type ServiceState struct {
+	// ChainID is the expected value of the `Result` field in any endpoint's response to an `eth_chainId` request.
+	ChainID string
+
+	Logger polylog.Logger
+
+	stateLock sync.RWMutex
+	// estimatedBlockNumber is the estimated current block number based on endpoints' responses to `eth_blockNumber` requests.
+	// It is calculated as the maximum of block height reported by any of the endpoints.
+	//
+	// See the following link for more details:
+	// https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_blocknumber
+	estimatedBlockNumber uint64
+}
+
+// TODO_FUTURE: add an endpoint ranking method which can be used to assign a rank/score to a valid endpoint to guide endpoint selection.
+//
+// ValidateEndpoint returns an error if the supplied endpoint is not valid based on the estimated state of the EVM blockchain.
+func (s *ServiceState) ValidateEndpoint(endpoint endpoint) error {
+	s.stateLock.RLock()
+	defer s.stateLock.RUnlock()
+
+	if err := endpoint.Validate(s.ChainID); err != nil {
+		return err
+	}
+
+	blockNumber, err := endpoint.GetBlockNumber()
+	if err != nil {
+		return err
+	}
+
+	if blockNumber < s.estimatedBlockNumber {
+		return fmt.Errorf("endpoint has block height %d, estimated block height is %d", blockNumber, s.estimatedBlockNumber)
+	}
```

Comment:
Can you move this to its own function called `validateEndpointBlockNumber` so we:
1. Have a foundation in place to add other check
2. Make it clear that block number is not the end all